File: 0107-binary-tree-level-order-traversal-ii/0107-binary-tree-level-order-traversal-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *   int val;
 *   TreeNode left;
 *   TreeNode right;
 *   TreeNode() {}
 *   TreeNode(int val) { this.val = val; }
 *   TreeNode(int val, TreeNode left, TreeNode right) {
 *     this.val = val;
 *     this.left = left;
 *     this.right = right;
 *   }
 * }
 */
class AuxNode {
  private int level;
  private TreeNode node;
  
  public AuxNode(int level, TreeNode node) {
    this.level = level;
    this.node = node;
  }
  
  public int getLevel() {
    return level;
  }
  
  public TreeNode getNode() {
    return node;
  }
}

class Solution {
  public List<List<Integer>> levelOrderBottom(TreeNode root) {
    List<AuxNode> queue = new ArrayList<AuxNode>();
    if (root != null) {
      AuxNode initial = new AuxNode(0, root);
      queue.add(initial);
    }
    List<List<Integer>> fw = new ArrayList();
    while (!queue.isEmpty()) {
      AuxNode curr = queue.remove(0);
      int level = curr.getLevel();
      TreeNode node = curr.getNode();
      if (level > fw.size() - 1) fw.add(new ArrayList<Integer>());
      List<Integer> currLevel = (List<Integer>) fw.get(level);
      currLevel.add(node.val);
      if (node.left != null) queue.add(new AuxNode(level + 1, node.left));
      if (node.right != null) queue.add(new AuxNode(level + 1, node.right));
    }
    Collections.reverse(fw);
    return fw;
  }
}

File: 0107-binary-tree-level-order-traversal-ii/0107-binary-tree-level-order-traversal-ii.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *   Val int
 *   Left *TreeNode
 *   Right *TreeNode
 * }
 */
type AuxNode struct {
  Level int
  Node *TreeNode
}

func levelOrderBottom(root *TreeNode) [][]int {
  queue := []*AuxNode{}
  if root != nil {
    queue = append(queue, &AuxNode{
      Level: 0,
      Node: root,
    })
  }
  fw := [][]int{}
  for len(queue) > 0 {
    curr := queue[0]
    queue = queue[1:]
    level, node := (*curr).Level, (*curr).Node
    if level > len(fw) - 1 {
      fw = append(fw, []int{})
    }
    fw[level] = append(fw[level], (*node).Val)
    if (*node).Left != nil {
      queue = append(queue, &AuxNode{
        Level: level + 1,
        Node: (*node).Left,
      })
    }
    if (*node).Right != nil {
      queue = append(queue, &AuxNode{
        Level: level + 1,
        Node: (*node).Right,
      })
    }
  }
  rv := [][]int{}
  for i := len(fw) - 1; i >= 0; i-- {
    rv = append(rv, fw[i])
  }
  return rv
}